### ISSUE TITLE
Distributed evaluation when metrics cannot be averaged across processes. 

### ISSUE THREAD

**Issue by dhruvdcoder** [allenai/allennlp#4050]
**System (please complete the following information):**
 - OS: All
 - Python version: Any 
 - AllenNLP version: Master
 - PyTorch version: 1.4

**Question**

Shouldn't the aggregation of evaluation metrics in the case of distributed evaluation be delegated to the concrete metric classes, instead of taking an average of the metrics across processes? This might not be a problem for metrics like accuracy but for something like recall or precision this can lead to very different numbers in single process evaluation vs distributed evaluation. In particular, I am talking about the following line in code:

[https://github.com/allenai/allennlp/blob/c497103542096037dc21c24466127bc6c820e75b/allennlp/training/util.py#L351](https://github.com/allenai/allennlp/blob/c497103542096037dc21c24466127bc6c820e75b/allennlp/training/util.py#L351)

Am I missing something? 

Thanks!

**Comment by mateuszpieniak** [comment]
I think that it is related. I try to reproduce MNLI results using ALBERT & distributed training using the Siamese-BERT model. Accuracy is my evaluation metric. It doesn't work with distributed code but seems to work using a single GPU. The "effective" hyperparameters are the same.

For distributed training:

```
{
...
   data_loader: {
     batch_sampler: {
       type: "bucket",
       batch_size : 2,
       sorting_keys: ['premise', 'hypothesis']
     }
   },
   distributed: if distributed then {
     num_nodes: 1,
     cuda_devices: [0, 1, 2, 3],
   },
   trainer: {
   ...
     num_gradient_accumulation_steps: 16,
   ...
   }
}
```

For single GPU:
```
{
...
   data_loader: {
     batch_sampler: {
       type: "bucket",
       batch_size : 2,
       sorting_keys: ['premise', 'hypothesis']
     }
   },
   trainer: {
   ...
     cuda_device: 0
     num_gradient_accumulation_steps: 64,
   ...
   }
}
```

![image](https://user-images.githubusercontent.com/31375424/83864595-3d29ce80-a725-11ea-9002-31fa032eac83.png)

**Comment by matt-gardner** [comment]
@dhruvdcoder, sorry that it looks like somehow missed this issue originally.

@epwalsh, I haven't ever touched the distributed code, but you've been doing a bunch with this recently.  Care to take a look at this issue?

**Comment by epwalsh** [comment]
Hey @mateuszpieniak, in those TensorBoard screenshots, is the orange line the distributed experiment and the red line the single GPU experiment?

**Comment by mateuszpieniak** [comment]
@epwalsh Yes. I think you might try to use config files from `allennlp-models` (https://github.com/allenai/allennlp-models/blob/master/training_config/pair_classification/mnli_roberta.jsonnet) to reproduce the bug I encountered.   

My example differs a bit from the mentioned config though.
- I don't combine `premise` and `hypothesis` using `[SEP]`, but I share the architecture for each input. as described in `Siamese-BERT` paper (https://arxiv.org/abs/1908.10084).  
- I don't use any feed-forward layer in the `basic_classifier`.
- I use ALBERTv2 architecture.

**Comment by dirkgr** [comment]
The issue title says this is about metrics that can't be aggregated by summing or averaging them, but those Tensorboard curves show that the loss doesn't improve either, and that has nothing to do with metrics. So what's the real issue? Is it that distributed training just doesn't work in this case?

**Comment by ZhaofengWu** [comment]
I think there're two separate issues discussed here. (1) loss not converging under DDP; (2) how should metrics be aggregated under DDP (the original question).

**Comment by dirkgr** [comment]
I think @epwalsh is working on several things involving distributed training already, so I'll declare that this issue is just about aggregating metrics properly.

**Comment by epwalsh** [comment]
Yeup, actually both of these issues are on my road map to investigate.

**Comment by epwalsh** [comment]
But if someone else wants to jump in, feel free.
